src/fitutil/SystematicManager.cpp

```diff
@@ -1,39 +1,196 @@
+
+        for (int j = 0; j < group.size(); ++j) {
+            if(std::find(allname.begin(),allname.end(), group.at(j)->GetName() ) == allname.end())
+                allname.push_back(group.at(j)->GetName());
+            else
+                throw NotFoundError(Formatter()<<
+                        "SystematicManager :: Systematic: "<< 
+                        group.at(i)->GetName()<<
+                        " is in more than one group. BAD!!!");
+        } 
+    }
 }
-                                        
-void 
-SystematicManager::Add(Systematic* systematic_){    
-    fSystematics.push_back(systematic_);
-    fNSystematics++;
+
+void
+SystematicManager::AddDist(const BinnedED& pdf, const std::vector<std::string>& syss_){
+    //Check whether all systematics in syss_ are unique.
+    UniqueSystematics(syss_);
+    fEDGroups[pdf.GetName()] = syss_;
+}
+void
+SystematicManager::AddDist(const BinnedED& pdf, const std::string& syss_){
+    AddDist(pdf,std::vector<std::string>(1,syss_));
+}
+
+void
+SystematicManager::DistortEDs(std::vector<BinnedED>& fWorkingEDs_) const {
+    for(size_t j = 0; j < fWorkingEDs_.size(); j++){
+        const std::string name = fWorkingEDs_.at(j).GetName();
+
+        //If default group exist then apply that set of systematics first regardless of name.
+        if ( fGroups.find("default") != fGroups.end() )
+            fWorkingEDs_[j].SetBinContents(GetTotalResponse("default").operator()(fWorkingEDs_.at(j).GetBinContents()));
+
+        if(fEDGroups.find(name) == fEDGroups.end())
+            continue;
+            
+        //Apply everything else.
+        for (int i = 0; i < fEDGroups.at(name).size(); ++i) {
+            //Check that the group has systematics init. 
+            std::string groupName = fEDGroups.at(name).at(i);
+            if (groupName == "default")
+                continue;
+            if (!fGroups.at( groupName ).size())
+                throw LogicError(Formatter()<<"SystematicManager:: ED "<<
+                        name
+                        <<" has a systematic group of zero size acting on it");
+
+            fWorkingEDs_[j].SetBinContents(GetTotalResponse(groupName).operator()(fWorkingEDs_.at(j).GetBinContents()));
+        }
+    }
 }
 
-size_t
+//Getters and Setters
+
+const size_t
+SystematicManager::GetNSystematicsInGroup(const std::string& name_) const{
+    try{        
+        return fGroups.find(name_)->second.size();
+    }
+    catch(const std::out_of_range& e_){
+        throw NotFoundError(Formatter()<<
+                "SystematicManager :: name : "<< 
+                name_ <<
+                " not found in systematic map");
+    }
+}
+
+const std::vector<std::string>
+SystematicManager::GetGroupNames() const{
+    std::vector<std::string> v;
+    for(std::map<std::string,std::vector<Systematic*> >::const_iterator it = fGroups.begin(); it !=fGroups.end(); ++it) {
+        v.push_back(it->first);
+    }
+    return v;
```

Comment:
Just FYI could be `return ContainerTools::GetKeys(fGroups);` if you like a shortcut